File: media_utils.py
"""
Media utilities for Google Drive URL handling and media type detection.
"""

import re
from typing import Optional

MAX_FILE_SIZE = 50 * 1024 * 1024  # 50 MB

GDRIVE_PATTERNS = [
    re.compile(r"drive\.google\.com/file/d/([a-zA-Z0-9_-]+)"),
    re.compile(r"drive\.google\.com/open\?id=([a-zA-Z0-9_-]+)"),
    re.compile(r"drive\.usercontent\.google\.com/download\?id=([a-zA-Z0-9_-]+)"),
    re.compile(r"drive\.google\.com/uc\?.*id=([a-zA-Z0-9_-]+)"),
]

IMAGE_CONTENT_TYPES = {
    "image/jpeg", "image/png", "image/gif", "image/webp", "image/bmp",
}
VIDEO_CONTENT_TYPES = {
    "video/mp4", "video/quicktime", "video/x-msvideo", "video/webm",
    "video/mpeg", "video/x-matroska",
}
SUPPORTED_CONTENT_TYPES = IMAGE_CONTENT_TYPES | VIDEO_CONTENT_TYPES

IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".gif", ".webp", ".bmp"}
VIDEO_EXTENSIONS = {".mp4", ".mov", ".avi", ".webm", ".mpeg", ".mpg", ".mkv"}


def is_google_drive_url(url: str) -> bool:
    """Check if a URL is a Google Drive link."""
    return extract_gdrive_file_id(url) is not None


def extract_gdrive_file_id(url: str) -> Optional[str]:
    """Extract the file ID from a Google Drive URL. Returns None if not a Drive URL."""
    for pattern in GDRIVE_PATTERNS:
        match = pattern.search(url)
        if match:
            return match.group(1)
    return None


def detect_media_type(content_type: Optional[str], filename: Optional[str]) -> str:
    """Detect whether media is PHOTO or VIDEO based on content-type or filename.

    Returns "PHOTO" or "VIDEO". Defaults to "PHOTO" if indeterminate.
    """
    if content_type:
        ct = content_type.split(";")[0].strip().lower()
        if ct in VIDEO_CONTENT_TYPES:
            return "VIDEO"
        if ct in IMAGE_CONTENT_TYPES:
            return "PHOTO"

    if filename:
        ext = _get_extension(filename)
        if ext in VIDEO_EXTENSIONS:
            return "VIDEO"
        if ext in IMAGE_EXTENSIONS:
            return "PHOTO"

    return "PHOTO"


def _get_extension(filename: str) -> str:
    """Extract lowercase file extension from a filename."""
    dot_idx = filename.rfind(".")
    if dot_idx == -1:
        return ""
    return filename[dot_idx:].lower()

File: node/src/media-utils.ts
/**
 * Media utilities for Google Drive URL handling and media type detection.
 */

export const MAX_FILE_SIZE = 50 * 1024 * 1024; // 50 MB

const GDRIVE_PATTERNS: RegExp[] = [
  /drive\.google\.com\/file\/d\/([a-zA-Z0-9_-]+)/,
  /drive\.google\.com\/open\?id=([a-zA-Z0-9_-]+)/,
  /drive\.usercontent\.google\.com\/download\?id=([a-zA-Z0-9_-]+)/,
  /drive\.google\.com\/uc\?.*id=([a-zA-Z0-9_-]+)/,
];

const IMAGE_CONTENT_TYPES = new Set([
  "image/jpeg",
  "image/png",
  "image/gif",
  "image/webp",
  "image/bmp",
]);

const VIDEO_CONTENT_TYPES = new Set([
  "video/mp4",
  "video/quicktime",
  "video/x-msvideo",
  "video/webm",
  "video/mpeg",
  "video/x-matroska",
]);

const IMAGE_EXTENSIONS = new Set([
  ".jpg",
  ".jpeg",
  ".png",
  ".gif",
  ".webp",
  ".bmp",
]);

const VIDEO_EXTENSIONS = new Set([
  ".mp4",
  ".mov",
  ".avi",
  ".webm",
  ".mpeg",
  ".mpg",
  ".mkv",
]);

/**
 * Check if a URL is a Google Drive link.
 */
export function isGoogleDriveUrl(url: string): boolean {
  return extractGdriveFileId(url) !== null;
}

/**
 * Extract the file ID from a Google Drive URL. Returns null if not a Drive URL.
 */
export function extractGdriveFileId(url: string): string | null {
  for (const pattern of GDRIVE_PATTERNS) {
    const match = url.match(pattern);
    if (match) {
      return match[1];
    }
  }
  return null;
}

/**
 * Detect whether media is PHOTO or VIDEO based on content-type or filename.
 * Returns "PHOTO" or "VIDEO". Defaults to "PHOTO" if indeterminate.
 */
export function detectMediaType(
  contentType: string | null,
  filename: string | null
): string {
  if (contentType) {
    const ct = contentType.split(";")[0].trim().toLowerCase();
    if (VIDEO_CONTENT_TYPES.has(ct)) return "VIDEO";
    if (IMAGE_CONTENT_TYPES.has(ct)) return "PHOTO";
  }

  if (filename) {
    const ext = getExtension(filename);
    if (VIDEO_EXTENSIONS.has(ext)) return "VIDEO";
    if (IMAGE_EXTENSIONS.has(ext)) return "PHOTO";
  }

  return "PHOTO";
}

function getExtension(filename: string): string {
  const dotIdx = filename.lastIndexOf(".");
  if (dotIdx === -1) return "";
  return filename.slice(dotIdx).toLowerCase();
}
